src/tcpip/tcpip.c

```diff
@@ -790,9 +746,17 @@ static void mg_tcpip_poll(struct mg_tcpip_if *ifp, uint64_t uptime_ms) {
   if (ifp->ip == 0 && expired_1000ms) tx_dhcp_discover(ifp);
 
   // Read data from the network
-  size_t len = ifp->driver->rx((void *) ifp->rx.ptr, ifp->rx.len, ifp);
-  if (len) {
-    mg_tcpip_rx(ifp, (void *) ifp->rx.ptr, len);
+  if (ifp->driver->rx != NULL) {  // Polling driver. We must call it
+    size_t len =
+        ifp->driver->rx(ifp->recv_queue.buf, ifp->recv_queue.size, ifp);
+    if (len > 0) mg_tcpip_rx(ifp, ifp->recv_queue.buf, len);
+  } else {  // Interrupt-based driver. Fills recv queue itself
+    char *buf;
+    size_t len = mg_queue_next(&ifp->recv_queue, &buf);
+    if (len > 0) {
+      mg_tcpip_rx(ifp, buf, len);
+      mg_queue_del(&ifp->recv_queue, len);
+    }
```

Comment:
We agreed on providing a single entry point and always read from rx(); the driver would pull from the queue if it pushed to it (IRQ drivers and RP2040 RMII example)
When introducing the RP2040 RMII example, we decided to change the API as the new one was more homogeneous, flexible, and consistent

Reply:
This is done by the IRQ-driven drivers using the queue by calling the function removed above. This way, IRQ or no IRQ, queue or no queue, the API stays the same, we call the driver rx() function.
Plus, drivers that don't use the queue (RP2040 WiFi IIRC) don't waste memory by having a queue they don't use.